subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/JdkIllegalReflectionTestWorkerIntegrationTest.groovy

```diff
@@ -0,0 +1,152 @@
+
+    def setup() {
+        buildFile << """
+            plugins {
+                id 'application'
+            }
+
+            ${mavenCentralRepository()}
+
+            dependencies {
+                testImplementation 'junit:junit:4.13'
+            }
+
+            application {
+                mainClass = 'example.Main'
+            }
+
+            test {
+                testLogging {
+                    showStandardStreams = true
+                }
+            }
+        """
+
+        file("src/main/java/example/Main.java") << """
+            package example;
+
+            import java.lang.invoke.MethodHandles;
+
+            public class Main {
+                public static void main(String[] args) {
+                    reflectUpon(java.lang.ClassLoader.class);
+                    reflectUpon(java.util.Random.class);
+                }
+
+                public static void reflectUpon(Class<?> clazz) {
+                    try {
+                        MethodHandles.privateLookupIn(clazz, MethodHandles.lookup());
+                        System.out.println("Success: " + clazz.getName());
+                    } catch (Exception e) {
+                        System.out.println(String.format("Failure: %s, %s", clazz.getName(), e.getMessage()));
+                    }
+                }
+            }
+        """
+
+        file("src/test/java/example/MainTest.java") << """
+            package example;
+
+            import org.junit.Test;
+
+            public class MainTest {
+                @Test
+                public void runTest() {
+                    Main.main(new String[0]);
+                }
+            }
+        """
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/19771")
+    def "#command command #success when application uses reflection on java #version "() {
+
+        given:
+        withVersion(version)
+
+        when:
+        run command
+
+        then:
+        if (success == "succeeds") {
+            outputContains "Success: java.lang.ClassLoader"
+            outputContains "Success: java.util.Random"
+        } else {
+            outputContains "Failure: java.lang.ClassLoader, module java.base does not open java.lang to unnamed module"
+            outputContains "Failure: java.util.Random, module java.base does not open java.util to unnamed module"
+        }
+
+        where:
+        command | version                | success
+        "test"  | JavaVersion.VERSION_17 | "fails"
+        "run"   | JavaVersion.VERSION_17 | "fails"
+        "test"  | JavaVersion.VERSION_16 | "fails"
+        "run"   | JavaVersion.VERSION_16 | "fails"
+        "test"  | JavaVersion.VERSION_15 | "succeeds"
+        "run"   | JavaVersion.VERSION_15 | "succeeds"
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/19771")
+    def "can add --add-opens flag in test for package #pkg to permit reflection on java #version"() {
+
+        given:
+        withVersion(version)
```

Comment:
I think we don't particularly care about the exact version of this except that it's new enough.

You can also get rid of using an exact version of the JDK by annotating this test with `@Requires(TestPrecondition.JDK16_OR_LATER)`.  On CI, these tests will only run when this precondition is met.